torch/lib/c10d/reducer.cpp

```diff
@@ -455,31 +520,38 @@ void Reducer::autograd_hook(VariableIndex index) {
     local_used_maps_[index.replica_index][index.variable_index] = 1;
   }
 
-  // Rebuild bucket only if 1) it is the first time to rebuild bucket 2)
-  // find_unused_parameters_ is false, currently it does not support when there
-  // are unused parameters 3) this backward pass needs to run allreduce. Here,
-  // we just dump tensors and their parameter indices into rebuilt_params_ and
-  // rebuilt_param_indices_ based on gradient arriving order, and then at the
-  // end of finalize_backward(), buckets will be rebuilt based on
-  // rebuilt_params_ and rebuilt_param_indices_, and then will be broadcasted
-  // and initialized. Also we only need to dump tensors and parameter indices of
-  // one replica.
-  push_rebuilt_params(index);
+  if (static_graph_ && num_iterations_ == 1) {
+    numGradHooksTriggeredMap_[index] += 1;
+    return;
+  }
 
   // If `find_unused_parameters_` is true there may be model parameters that
   // went unused when computing the model output, they won't be part of the
   // autograd graph, and won't receive gradients. These parameters are
   // discovered in the `prepare_for_backward` function and their indexes stored
   // in the `unused_parameters_` vector.
-  if (!has_marked_unused_parameters_ && find_unused_parameters_) {
+  if (!has_marked_unused_parameters_) {
     has_marked_unused_parameters_ = true;
     for (const auto& unused_index : unused_parameters_) {
       mark_variable_ready(unused_index);
     }
   }
 
-  // Finally mark variable for which this function was originally called.
-  mark_variable_ready(index);
+  // If it is static graph, after 1st iteration, check a avariable
+  // is ready for communication based on numGradHooksTriggeredMap_.
+  if (static_graph_ && num_iterations_ > 1) {
+    TORCH_CHECK(
+        numGradHooksTriggeredMapPerIteration_[index] > 0,
+        "Your training graph is changed in this iteration, "
+        "try not to specify static_graph_ to be true in this case.");
```

Comment:
if the number never went to 0, it will not be marked as Ready for communication, then it will not trigger finalize_backward, and ensure_prior_reduction_finished() will trow errors in this case